File: frontend/src/types/relation.ts
export type RelationLineStyle = "solid" | "dashed" | "dotted";

export type RelationLabel = {
  id: string;
  project_id: string;
  name: string;
  emoji: string;
  color: string;
  line_style: RelationLineStyle;
  sort_order: number;
  created_at: string;
  updated_at: string;
};

export type Relation = {
  id: string;
  project_id: string;
  source_paper_id: string;
  target_paper_id: string;
  label_id: string;
  created_at: string;
  updated_at: string;
};

export type IncomingRelationGroup = {
  label_id: string;
  source_paper_ids: string[];
};

export type IncomingRelations = {
  target_paper_id: string;
  groups: IncomingRelationGroup[];
};

export type IncomingRelationsUpdate = {
  groups: IncomingRelationGroup[];
};

export type RelationLabelInput = {
  name: string;
  emoji: string;
  color: string;
  line_style: RelationLineStyle;
  sort_order: number;
};

File: backend/app/schemas/relation.py
from datetime import datetime

from pydantic import BaseModel, Field


class RelationLabelCreate(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    emoji: str = Field(default="🔗", max_length=16)
    color: str = "#b8653f"
    line_style: str = "solid"
    sort_order: int = 0


class RelationLabelUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    emoji: str = Field(default="🔗", max_length=16)
    color: str = "#b8653f"
    line_style: str = "solid"
    sort_order: int = 0


class RelationLabelOut(BaseModel):
    id: str
    project_id: str
    name: str
    emoji: str
    color: str
    line_style: str
    sort_order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RelationOut(BaseModel):
    id: str
    project_id: str
    source_paper_id: str
    target_paper_id: str
    label_id: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class IncomingRelationGroupIn(BaseModel):
    label_id: str
    source_paper_ids: list[str] = Field(default_factory=list)


class IncomingRelationsUpdate(BaseModel):
    groups: list[IncomingRelationGroupIn] = Field(default_factory=list)


class IncomingRelationGroupOut(BaseModel):
    label_id: str
    source_paper_ids: list[str]


class IncomingRelationsOut(BaseModel):
    target_paper_id: str
    groups: list[IncomingRelationGroupOut]
